ansible-galaxy collection - validate and refresh KeycloakToken

s-hertel opened the pull request:
##### SUMMARY
This should fix refreshing KeycloakTokens while waiting for import results. I'm not sure how to test it though...

Fixes #70019

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
ansible-galaxy collection publish

s-hertel committed:
Approach 1 (naive)

add expiration property to KeycloakToken

warn when token expires

handle expired token when waiting for imports to complete after publishing to AH

changelog

s-hertel commented on lib/ansible/galaxy/api.py:
```diff
@@ -696,17 +697,26 @@ def wait_import_task(self, task_id, timeout=0):
+                if e.http_code == 401 and self.token and isinstance(self.token, KeycloakToken):
+                    if token_expired:
+                        raise
+                    display.vvv("Authentication failed, refreshing token to try again.")
```
I'm going to make this a warning instead of debug msg.

jborean93 reviewed (approved):
+1 to the warning, the rest looks good.

webknjaz commented on lib/ansible/galaxy/api.py:
```diff
@@ -696,17 +697,26 @@ def wait_import_task(self, task_id, timeout=0):
                                          error_context_msg='Error when getting import task results at %s' % full_url)
             except GalaxyError as e:
-                if e.http_code != 404:
+                if e.http_code == 401 and self.token and isinstance(self.token, KeycloakToken):
```
Checking for `self.token` is probably redundant. It seems to be an `is None` check, but `KeycloakToken` doesn't even implement `__bool__` so that's kinda pointless. Or is it some sort of optimization?
```suggestion
                if e.http_code == 401 and isinstance(self.token, KeycloakToken):
```

s-hertel replied:
Yeah, it wasn't necessary, I just saw that `self.token` is initially None. But that will still work with the isinstance check.

webknjaz replied:
exactly

webknjaz commented on lib/ansible/galaxy/api.py:
```diff
@@ -696,17 +697,26 @@ def wait_import_task(self, task_id, timeout=0):
             except GalaxyError as e:
-                if e.http_code != 404:
+                if e.http_code == 401 and self.token and isinstance(self.token, KeycloakToken):
+                    if token_expired:
```
Is this a check just for the second failed iteration in a row? Maybe have a counter to represent this?

s-hertel replied:
I could make it a counter, but unless we attempt refreshing the token more than once in a row, I think the boolean reads a little better.

webknjaz replied:
Seems like with https://github.com/ansible/ansible/pull/79145/files/0c2d16d6eaa97b5e1dc95cbddb199acac5aa0d74#r1013488315 refreshing would be redundant.

s-hertel replied:
Yep, just haven't removed this yet. I'll request your review FYI when I've implemented your suggestions.

webknjaz commented on lib/ansible/galaxy/api.py:
```diff
@@ -696,17 +697,26 @@ def wait_import_task(self, task_id, timeout=0):
+                        raise
+                    display.vvv("Authentication failed, refreshing token to try again.")
+                    self.token._token = None
+                    token_expired = True
```
I think it'd be more organic to implement a `KeycloakToken.expired` property. The `KeycloakToken.get()` method would need to get updated to store the expiration time.
And then, there could be a `KeycloakToken.expire()` method for resetting the token.

I'd also recommend adding a skew offset to the `KeycloakToken.expired` property. So it'd "expire" in on the client side slightly earlier than it does on the server.
This could later be used in the API call wrapper to pre-refresh the token if it's about to expire. A refresh call could be added before this line https://github.com/ansible/ansible/blob/f02b8fb/lib/ansible/galaxy/api.py#L428 to update the token if it's turning stale.

This proactive approach would make the operation smoother than the reactive one, most of the time.

s-hertel replied:
Good idea. Rather than an `expire` method to update the token, `KeycloakToken.get()` handles the refresh for expired tokens. `KeycloakToken.headers()` uses `KeycloakToken.get()` as well, so this should all be implemented now.

webknjaz commented on lib/ansible/galaxy/token.py:
```diff
@@ -54,11 +59,19 @@ def __init__(self, access_token=None, auth_url=None, validate_certs=True, client
 
+    @property
+    def expired(self):
+        issue_time = datetime.strptime(self._issued_at, ISO_8601_FORMAT_STR)
```
I think Python 3 has functions that already use ISO 8601 format so no need for a constant: https://docs.python.org/3/library/datetime.html#datetime.datetime.isoformat. Would this work?

s-hertel replied:
Python-specific ISO 8601 doesn't seem to handle the timezone.
```python
>>> datetime.fromisoformat("2022-11-27T13:50:51.925209Z")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Invalid isoformat string: '2022-11-27T13:50:51.925209Z'
>>> datetime.strptime("2022-11-27T13:50:51.925209Z", "%Y-%m-%dT%H:%M:%S.%f%z").tzinfo
datetime.timezone.utc
>>> datetime.fromisoformat("2022-11-27T13:50:51.925209").tzinfo
>>>
```

webknjaz replied:
Hm... I seem to recall addressing this somehow. Let me see.

webknjaz replied:
Aha! `Z` is sometimes ambiguous: https://en.wikipedia.org/wiki/ISO_8601#Coordinated_Universal_Time_(UTC). Maybe that's why.

FWIW here's the solution I used in octomachinery: https://github.com/sanitizers/octomachinery/blob/94c3eaa7/octomachinery/github/models/utils.py#L27-L31.

webknjaz replied:
Specifically this works:
```pycon
In [9]: datetime.fromisoformat("2022-11-27T13:50:51.925209+00:00").tzinfo
Out[9]: datetime.timezone.utc
```

s-hertel replied:
Thanks for the pointer about Z, TIL. The format from AH is an integer timestamp which made this simpler.

s-hertel commented on lib/ansible/galaxy/token.py:
```diff
@@ -68,8 +68,16 @@ def __init__(self, access_token=None, auth_url=None, validate_certs=True, client
     def expired(self):
-        issue_time = datetime.strptime(self._issued_at, ISO_8601_FORMAT_STR)
-        almost_expiration = issue_time + timedelta(seconds=self._expires_in - 3)  # expire a few seconds before to handle proactively
+        if None in (self._issued_at, self._expires_in):
```
I probably should update the existing unit tests instead of this...

webknjaz replied:
Also, I'm not sure if the "contains" check uses the identity (`is`) or value comparison (`==`) to match the elements internally. So there might be some unexpected behavior hidden behind such a check.

webknjaz commented on lib/ansible/galaxy/token.py:
```diff
@@ -54,11 +59,27 @@ def __init__(self, access_token=None, auth_url=None, validate_certs=True, client
+            display.warning(f"KeycloakToken issued at unknown time format {self._issued_at}")
+            return False
+
+        # expire a few seconds before to handle proactively
```
I saw this value being a whole minute in a GitHub bot project that had to work with GitHub API. The reasoning was that it'd also accommodate for the system clock being slightly out of sync with the actual time (when NTP detects such anomalies, it tends to correct time gently with multiple iterations instead of jumping the clock for relatively big amounts of time).

What is the typical expiration time for Galaxy tokens? If it's about an hour, 60 seconds might be a good trade-off.

webknjaz replied:
Found where I saw that: https://github.com/python-trio/snekomatic/blob/43e8e6d/snekomatic/gh.py#L278 / https://github.com/python-trio/snekomatic/blob/43e8e6d/snekomatic/gh.py#L86-L93.

s-hertel replied:
At least in the example I saw, it was 300 seconds, but I have yet to test/verify that's the current behavior. Guessing it's not an hour though if people are running into the token expiring when publishing a collection...

webknjaz replied:
Ah, fair enough. Maybe we need to tune it to correlate with the slow cases of HTTP request timeouts?

s-hertel replied:
The expiration for an AH token is 15 minutes, so I've changed this to refresh the token a minute early. Seem okay to you?

webknjaz replied:
Yep, that should be more than enough.

webknjaz commented on lib/ansible/galaxy/token.py:
```diff
@@ -54,11 +59,27 @@ def __init__(self, access_token=None, auth_url=None, validate_certs=True, client
+            return False
+
+        # expire a few seconds before to handle proactively
+        almost_expiration = issue_time + timedelta(seconds=self._expires_in - 3)
```
Plz also put the numeric value in a well-readable constant.

s-hertel replied:
named `AH_TOKEN_EXPIRES_N_MIN_EARLY`, feel free to bikeshed

webknjaz commented on lib/ansible/galaxy/token.py:
```diff
@@ -87,8 +108,10 @@ def get(self):
-        # - extract 'access_token'
+        # - extract token details
         self._token = data.get('access_token')
+        self._issued_at = data.get('issued_at')
```
How about updating the condition at the beginning of this method too?
```diff
-if self._token:
+if self._token and not self.expired:
     return self._token
```

webknjaz commented on lib/ansible/galaxy/api.py:
```diff
@@ -696,17 +697,29 @@ def wait_import_task(self, task_id, timeout=0):
+                    if token_expired:
+                        raise
+                    display.vvv("Authentication failed, refreshing token to try again.")
+                    self.token._token = None
```
With https://github.com/ansible/ansible/pull/79145/files/0c2d16d6eaa97b5e1dc95cbddb199acac5aa0d74#r1013488315, it won't be necessary to mutate the private attributes of the token object.

webknjaz commented on lib/ansible/galaxy/api.py:
```diff
@@ -696,17 +697,29 @@ def wait_import_task(self, task_id, timeout=0):
         while timeout == 0 or (time.time() - start) < timeout:
+            if isinstance(self.token, KeycloakToken) and self.token.expired:
+                display.warning("Authentication failed, refreshing token to try again.")
+                self.token._token = None
```
With https://github.com/ansible/ansible/pull/79145/files/0c2d16d6eaa97b5e1dc95cbddb199acac5aa0d74#r1013488315, it won't be necessary to mutate the private attributes of the token object.

webknjaz commented on lib/ansible/galaxy/api.py:
```diff
@@ -696,17 +697,29 @@ def wait_import_task(self, task_id, timeout=0):
 
+        token_expired = False
         while timeout == 0 or (time.time() - start) < timeout:
+            if isinstance(self.token, KeycloakToken) and self.token.expired:
```
`isinstance()` is relatively expensive. How about moving the detection outside the loop and just reusing the result here?
```python
is_keycloak_token = isinstance(self.token, KeycloakToken)
```

```suggestion
            if is_keycloak_token and self.token.expired:
```

webknjaz commented on lib/ansible/galaxy/api.py:
```diff
@@ -696,17 +697,29 @@ def wait_import_task(self, task_id, timeout=0):
             except GalaxyError as e:
-                if e.http_code != 404:
+                if e.http_code == 401 and self.token and isinstance(self.token, KeycloakToken):
+                    if token_expired:
```
Can this be
```suggestion
                    if self.token.expired:
```
now?

webknjaz commented on test/units/galaxy/test_token.py:
```diff
@@ -96,3 +101,35 @@ def test_token_from_file_missing(b_token_file):
+    response2 = u'{{"access_token": "token", "issued_at": "{0}", "expires_in": "{1}"}}'.format(
+        (datetime.now(timezone.utc) - timedelta(seconds=298)).strftime(time_format), "300"
+    )
+    mock_open = MagicMock(side_effect=[StringIO(response1), StringIO(response2)])
```
FYI `StringIO` may leave unclosed resources/file handles in memory, which is why it's better to use `with StringIO(...) as ...:` or something like that. This sort of thing tends to become evident when executing code with warnings turned into errors. Like `python3 -Werror` or pytest's `filterwarnings = error` which is typically recommended to have set up by default.

s-hertel replied:
Fixed

s-hertel committed:
Approach 2 (naive, proactive)

Handle refreshing in KeycloakToken.get() before tokens expire

Make the warning a debug msg since there is no error

Use global constant for early KeycloakToken expiration

Co-authored-by: Sviatoslav Sydorenko <wk.cvs.github@sydorenko.org.ua>

webknjaz commented on lib/ansible/galaxy/token.py:
```diff
@@ -54,18 +58,26 @@ def __init__(self, access_token=None, auth_url=None, validate_certs=True, client
-        if self._token:
+        if self._token and not self.expired:
             return self._token
+        if self._token and self.expired:
```
The second clause is redundant, plus the log message gives it away. This is probably unnecessary.
```suggestion
        if self._token:
```

webknjaz commented on lib/ansible/galaxy/token.py:
```diff
@@ -86,9 +98,13 @@ def get(self):
         # TODO: handle auth errors
 
         data = json.loads(to_text(resp.read(), errors='surrogate_or_strict'))
+        token_data = json.loads(
+            base64.b64decode(data['id_token'].split('.')[1] + "==")
+        )
```
This incantation looks out of place. It leaks into a low-level abstraction of decoding and parsing the internals.

I'd move at least the last instruction into a pure function so it'd look something like
```python
bearer_token_struct = parse_bearer_token(resp.read())
```
It could return a data structure with both the access token and the id token, if it replaces both instructions.

I don't have visibility into what the HTTP response structure it, but often this data also includes some checksum/signature. Such a function could also verify the integrity and raise an error if it's corrupted somehow.

Maybe it could also normalize the time fields into datetime objects.

s-hertel replied:
Good suggestions, thanks. I implemented signature checking. I'm not sure I can assume the JSON web key sets url is relative to the auth URL, so I added a new server config option to achieve that in case the auth_url does not match AH - maybe not backportable now, but it's better I think.

I also fixed normalizing the datetime fields in parse_bearer_token.

Need to figure out how to get sanity tests passing, but I think this is close to done now.

webknjaz reviewed (approved):
Approving, but there's room for improvement if you're up to it :)

s-hertel committed:
Approach 3 (validate according to https://www.rfc-editor.org/rfc/rfc7519)

Add an openid_configuration galaxy server option. Since we don't want to
hardcode the AH URL, the bugfix is now opt-in only. On the plus side,
now the auth_url can be derived from the OpenID config, so users only
need to configure 1 of these options.

validate bearer token before using expiration data from it

normalize datetime fields before returning token data

add cryptograph to unit test requirements for ansible-galaxy

tweak user_agent string so getting AH OpenID config works (comment order
seems to have requirements?)

Co-authored-by: Sviatoslav Sydorenko <wk.cvs.github@sydorenko.org.ua>

s-hertel commented on lib/ansible/galaxy/token.py:
```diff
@@ -35,6 +42,89 @@
+    decoded_bytes = base64.urlsafe_b64decode(
+        to_bytes(data, errors="surrogate_or_strict") + b"=="
+    )
+    # print(f"{["%02x" % byte for byte in decoded_bytes]}")
```
```suggestion
```

webknjaz commented on lib/ansible/galaxy/token.py:
```diff
@@ -35,6 +48,93 @@
+        raise InvalidSignature(f"unsupported format: {data}")
+
+    if any(whitespace in data.rsplit('.')[0] for whitespace in string.whitespace):
+        raise InvalidSignature(f"JWS header and payload should not contain line breaks or whitespace: {data.split('.')[0]}")
```
Why does this use `str.split()` but the conditional has `str.rsplit()`?

s-hertel replied:
it was a typo, fixed

webknjaz commented on lib/ansible/galaxy/token.py:
```diff
@@ -35,6 +48,93 @@
+    if len(data.split('.')) != 3:
+        raise InvalidSignature(f"unsupported format: {data}")
+
+    if any(whitespace in data.rsplit('.')[0] for whitespace in string.whitespace):
```
`data.rsplit('.')[0]` is evaluated on each iteration and is probably not the cheapest operation here... Maybe reverse the order?

```suggestion
    if any(
        possible_whitespace in string.whitespace
        for possible_whitespace in data.rsplit('.')[0]
    ):
```

Also, can we check against the allowlist rather than a blocklist? Is there a spec for this? Is it mostly ASCII?

s-hertel replied:
Fixed.

Yeah, it's mostly ascii https://datatracker.ietf.org/doc/html/rfc4648#section-5, but we already validate it by decoding it,  so I don't think it's necessary to hardcode a valid character set. Whitespace is invalid (validation steps: https://www.rfc-editor.org/rfc/rfc7515#section-5.2), but base64.urlsafe_b64decode strips it out, so I added this check first.

webknjaz commented on lib/ansible/galaxy/token.py:
```diff
@@ -35,6 +48,93 @@
+
+
+def validate_jwt_format(data):
+    if len(data.split('.')) != 3:
```
Creating an in-memory list only to count its elements is probably more expensive than this:
```suggestion
    if data.count('.') != 3:
```

webknjaz commented on lib/ansible/galaxy/token.py:
```diff
@@ -35,6 +48,93 @@
+PublicNumbers = namedtuple("PublicNumbers", ["e", "n"])
+
+
+def validate_jwt_format(data):
```
How about improving this signature?
```suggestion
def assert_jwt_structure(raw_jwt_string):
    """Test that given JWT string looks valid.

    :raises InvalidSignature: If it isn't.
    """
```

s-hertel replied:
Added, also renamed and added the doc string for the header validation function.

webknjaz commented on lib/ansible/galaxy/token.py:
```diff
@@ -35,6 +48,93 @@
+    validate_jwt_format(data)
+
+    data = to_bytes(data)
+    signing_input = data.rsplit(b".", 1)[0]
+    b64header_data, b64payload_data, b64signature = data.split(b".")
```
Could this be
```suggestion
    signing_input, b64signature = data.rsplit(b".", 1)
    b64header_data, b64payload_data = signing_input.split(b".")
```
? Or am I missing something?

s-hertel replied:
It's equivalent, it doesn't achieves anything different?

webknjaz replied:
Yep, the main motivation was readability and semantical step separation, IIRC. Less chaining == easier to follow.

webknjaz commented on lib/ansible/galaxy/token.py:
```diff
@@ -35,6 +48,93 @@
+    signing_input = data.rsplit(b".", 1)[0]
+    b64header_data, b64payload_data, b64signature = data.split(b".")
+
+    header = json.loads(base64.urlsafe_b64decode(b64header_data + b"=="))
```
I see this sort of conversion both here and on the lower abstraction layer (in `b64url_to_long()`). Maybe, this warrants a helper?

s-hertel replied:
Ok, done.

webknjaz commented on lib/ansible/galaxy/token.py:
```diff
@@ -35,6 +48,93 @@
+
+    if public_numbers is None:
+        for key in jwks:
+            if key['kid'] == header['kid']:
```
Instead of accessing `header['kid']` on each iteration, maybe name it before the loop and reference the var?

s-hertel replied:
done

webknjaz commented on lib/ansible/galaxy/token.py:
```diff
@@ -35,6 +48,93 @@
+    token_info = json.loads(base64.urlsafe_b64decode(b64payload_data + b"=="))
+    signature = base64.urlsafe_b64decode(b64signature + b"==")
+
+    if public_numbers is None:
+        for key in jwks:
+            if key['kid'] == header['kid']:
+                e = key['e']
+                n = key['n']
+                break
+        else:
+            raise InvalidSignature(f"Public key {header['kid']} was not found in JWKS")
+    else:
+        e, n = public_numbers
```
It'd be cleaner to move the extraction into a pure function + label `e` and `n` with some actual names.

webknjaz replied:
Maybe even make it return an instance of `RSAPublicNumbers`.

s-hertel replied:
This was here to handle a test case, but I've removed it (and PublicNumbers) and refactored this into a function so I could just monkeypatch the public key numbers to use the test key values.

webknjaz commented on lib/ansible/galaxy/token.py:
```diff
@@ -35,6 +48,93 @@
 
 display = Display()
 
+AH_TOKEN_EXPIRES_N_MIN_EARLY = 1
+AH_AUTH_URL = "https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/token"
+AH_JWKS_URI = "https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/certs"
```
I'd expect these to be in the constants/config module, so it's possible for the users to override them cleanly.

s-hertel replied:
We don't configure AH by default, just document the auth_url `https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/token`. If I set the AH JWKS in config, it would break non-AH servers using KeycloakToken.

I'd like to fix AH tokens expiring on import without requiring people change anything, but I'm not sure how to do that without hardcoding. I wish `auth_url` had been an option for the OpenID configuration instead (https://sso.redhat.com/auth/realms/redhat-external/.well-known/openid-configuration), so the `token_endpoint` and `jkws_uri` could be retrieved from the same source. I wonder if that would be reasonable? (fix would still require people to change something, but wouldn't require 2 options to do the same task)

webknjaz replied:
If they are coupled, this sounds like a good idea.

webknjaz commented on lib/ansible/galaxy/token.py:
```diff
@@ -35,6 +48,93 @@
+AH_JWKS_URI = "https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/certs"
+
+# https://www.rfc-editor.org/rfc/rfc7518#section-3.1
+REQUIRED_JWT_HEADER_KEYS = set(['alg', 'typ', 'kid'])
```
Unless this module needs to be compatible with Python 2.6, we should really just use native syntax:
```suggestion
REQUIRED_JWT_HEADER_KEYS = {'alg', 'typ', 'kid'}
```

webknjaz commented on lib/ansible/galaxy/token.py:
```diff
@@ -35,6 +48,93 @@
+    'typ': ['JWT'],
+}
+
+PublicNumbers = namedtuple("PublicNumbers", ["e", "n"])
```
Is this unused in this module? I only see a single use in the test module. If it's really needed, maybe it should just be in tests?

s-hertel replied:
I refactored so I could use test key values more easily and removed this.

webknjaz commented on lib/ansible/galaxy/token.py:
```diff
@@ -35,6 +48,93 @@
+        if not e.args:
+            # give some debug info for clues
+            display.vvvv(f"{signing_input} could not be verified with signature {signature}")
+            display.vvvv(f"Public key exponent: {e}")
+            display.vvvv(f"Public key modulus: {n}")
```
Best just access the actual values passed.
```suggestion
            display.vvvv(f"Public key exponent: {public_key.e}")
            display.vvvv(f"Public key modulus: {public_key.n}")
```

s-hertel replied:
I used the values from the server intentionally,  so a user can easily identify if the correct key was used (which could indicate a bug in `ansible-galaxy`).  If I use the public_key attributes, the user is going to have to reverse engineer the long to see which public key was used.

webknjaz commented on lib/ansible/galaxy/token.py:
```diff
@@ -87,9 +199,31 @@ def get(self):
 
-        # - extract 'access_token'
+        if self.jwks is None and self._jwks_url is not None:
+            jwks_resp = open_url(
```
Can this be a CM? I see `jwks_resp.read()` but not `jwks_resp.close()` meaning that we'll probably have a resource leak here.

s-hertel replied:
Good catch, fixed.

webknjaz commented on lib/ansible/galaxy/token.py:
```diff
@@ -87,9 +199,31 @@ def get(self):
+        # - extract token details
         self._token = data.get('access_token')
 
+        if self.jwks is not None and HAS_CRYPTO:
```
Why not fail on no crypto? I'd expect it to be mandatory for this to work...

s-hertel replied:
This was for backwards compatibility since the token is used without validation on devel. Failing on no crypto is maybe fine since it's required for ansible-core, but it wasn't necessary before for ansible-galaxy (similarly, I started using `packaging` recently in `ansible-galaxy` but it caused some inconvenience even though it was already a core dep https://github.com/ansible/ansible/issues/78275). So if I'm doing that, I'd want to make a note of it in the changelog. As it is now, only the fix is dependent on crypto, and without it the behavior will stay the same (i.e. the token will be used until it expires).

webknjaz commented on lib/ansible/galaxy/token.py:
```diff
@@ -87,9 +199,31 @@ def get(self):
+        # - extract token details
         self._token = data.get('access_token')
 
+        if self.jwks is not None and HAS_CRYPTO:
+            try:
+                token_data = parse_bearer_token(data['id_token'], jwks=self.jwks)
+            except InvalidSignature as e:
+                # something went wrong parsing or validating the JWT, this is unexpected
+                # might need to support other algorithms?
+                # this isn't fatal to keep backwards compat, just won't get to use proactive token refreshing
+                orig_error = f': {e}' if e.args else ''
+                display.warning(f'Failed to verify JWS{orig_error}. The token will not be refreshed during long-running operations.')
+            else:
+                self._expires_at = token_data.get('exp')
```
Seems like this block could be moved to a helper function.

s-hertel replied:
The current code feels readable to me and since nothing is repeated I don't think a helper function seems useful here.

webknjaz commented on test/units/galaxy/test_token.py:
```diff
@@ -6,16 +6,110 @@
+    }
+
+
+@pytest.fixture()
```
nit
```suggestion
@pytest.fixture
```

webknjaz commented on test/units/galaxy/test_token.py:
```diff
@@ -6,16 +6,110 @@
+    iqmp = b64url_to_long(private_key_data['qi'])
+
+    private_numbers = RSAPrivateNumbers(p, q, d, dmp1, dmq1, iqmp, public_numbers)
+    yield private_numbers.private_key()
```
There's no need for a generator unless you do some clean-up post-`yield`.
```suggestion
    return private_numbers.private_key()
```

webknjaz commented on test/units/galaxy/test_token.py:
```diff
@@ -96,3 +190,51 @@ def test_token_from_file_missing(b_token_file):
+
+
+def test_parse_bearer_token(monkeypatch):
+    monkeypatch.setattr(token_file, 'REQUIRED_JWT_HEADER_KEYS', set(['alg']))
```
```suggestion
    monkeypatch.setattr(token_file, 'REQUIRED_JWT_HEADER_KEYS', {'alg'})
```

webknjaz commented on test/units/galaxy/test_token.py:
```diff
@@ -96,3 +190,51 @@ def test_token_from_file_missing(b_token_file):
+    public_numbers = token_file.PublicNumbers(n=n, e=e)
+    token_file.parse_bearer_token(token_info, public_numbers=public_numbers)
+
+    with pytest.raises(token_file.InvalidSignature):
```
It is the best practice to always use `pytest.raises()` with the `match=` argument. Otherwise, expect the future false-positives no go undetected.

s-hertel replied:
There is no message but I didn't think our tests should be asserting that (if cryptography adds a message, our tests would still be valid). I checked the cryptography unit tests, and they use this pattern without `match`.

webknjaz replied:
Okay, I missed the fact that this exception isn't ours. Though, if there are different situations when a generic exception may occur, it's advised to check the message strictly. If cryptography were to add several different messages for the exception, it'd be useful for Us to be notified about that since we probably want one specific case to happen predictably.

s-hertel replied:
Yes, of course.

webknjaz commented on test/units/galaxy/test_token.py:
```diff
@@ -96,3 +190,51 @@ def test_token_from_file_missing(b_token_file):
+
+@pytest.mark.parametrize('exp_delta', [1, -15, 15])
+def test_keycloak_expiration(exp_delta, monkeypatch, private_key, public_key_numbers):
+    expected_expired_result = exp_delta <= 1
```
I'd put this in the matrix to set clear expectations. This is a piece of logic that is not really what we test here, and such things only complicate how the test is perceived.

s-hertel replied:
fixed

webknjaz commented on test/units/galaxy/test_token.py:
```diff
@@ -96,3 +190,51 @@ def test_token_from_file_missing(b_token_file):
+def test_keycloak_expiration(exp_delta, monkeypatch, private_key, public_key_numbers):
+    expected_expired_result = exp_delta <= 1
+
+    exp_timestamp = int((datetime.now(tz=timezone.utc) + timedelta(minutes=exp_delta)).timestamp())
+    signed_payload = create_token_payload(private_key, exp_timestamp)
+
+    token = KeycloakToken(access_token='', auth_url='')
+    token.jwks = [{'kid': 'key_id', 'e': public_key_numbers['e'], 'n': public_key_numbers['n']}]
```
Maybe this test setup stage should be moved to fixtures?

s-hertel replied:
Moving 2 lines to a helper function that isn't generic or used by any other tests seems like overkill.

webknjaz commented on test/units/galaxy/test_token.py:
```diff
@@ -96,3 +190,51 @@ def test_token_from_file_missing(b_token_file):
+
+    mock_response = json.dumps({"access_token": "token", "id_token": signed_payload})
+    with StringIO(mock_response) as response:
+        monkeypatch.setattr(token_file, 'open_url', MagicMock(return_value=response))
```
pro tip: we have `pytest-mock` integrated, meaning that we can inject the `mocker` fixture and access this through `mocker.MagicMock` instead of relying on the imports external to the pytest machinery.

s-hertel replied:
Ok, I removed the existing import and updated the whole file.

webknjaz commented on lib/ansible/galaxy/token.py:
```diff
@@ -35,6 +47,110 @@
+    return int(decoded_bytes.hex(), 16)
+
+
+def b64urlsafe_to_json(data):
```
```suggestion
def b64urlsafe_to_dict(data):
```
Reading this, I don't think it returns a JSON string tho.

webknjaz replied:
Or
```suggestion
def parse_b64urlsafe_as_json(data):
```

webknjaz commented on lib/ansible/galaxy/token.py:
```diff
@@ -35,6 +47,110 @@
+    signature = base64.urlsafe_b64decode(b64signature + b"==")
+
+    public_key_numbers = get_public_numbers(header, jwks)
+    if public_key_numbers is None:
```
It took me a minute to realize when `get_public_numbers()` returns `None` since it's implicit. I think that rasing the exception belongs at the end of that helper function so that there's no need for the calling side to remember to make a check for an implicit magic value with special meaning. After all, there's no other input validation in it.

webknjaz commented on test/units/galaxy/test_token.py:
```diff
@@ -96,3 +189,50 @@ def test_token_from_file_missing(b_token_file):
+        token_file.parse_bearer_token(token_info + 'wrongsig')
+
+
+@pytest.mark.parametrize('exp_delta, expired', [(1, True), (-15, True), (15, False)])
```
Nit
```suggestion
@pytest.mark.parametrize(('exp_delta', 'expired'), [(1, True), (-15, True), (15, False)])
```

webknjaz reviewed (commented):
Got a few more comments...

s-hertel committed:
rebase fail, no docs
